### PR TITLE
implement a modular structure for metadata backends

### PR DESCRIPTION
**Pull request by zoltan**
currently the foundationdb backend if built as if it were a real shared library. however, this can only be used from within hyperwarp, and is not reusable (being tightly coupled to metadata itself), while we want our metadata library to be actually not coupled to any particular backend.

this change implements this idea.

### CODE REVIEW

**Comment by zoltan** [comment]
rebased on latest master

File: metadata-foundationdb/metadata-foundationdb.h
Comment (nielspardon) [inline metadata-foundationdb/metadata-foundationdb.h]: you renamed that to `load_metadata()` above

File: samples/metadata-sample.c
Comment (nielspardon) [inline samples/metadata-sample.c]: can you add the new way of persisting metadata here and not just comment out the old way?

File: samples/metadata-sample.c
Comment (nielspardon) [inline samples/metadata-sample.c]: can you add the new way of loading metadata here and not just comment out the old way?

File: metadata-foundationdb/metadata-foundationdb.c
Comment (nielspardon) [inline metadata-foundationdb/metadata-foundationdb.c]: I think we also need a `foundationdb_backend_finalize` in order to destroy the database and stop the network thread.

File: samples/metadata-sample.c
Comment (nielspardon) [inline samples/metadata-sample.c]: I think this should be uncommented again.

File: samples/metadata-sample.c
Comment (nielspardon) [inline samples/metadata-sample.c]: This should be implemented by the `foundationdb_backend_finalize` mentioned above.

File: metadata-foundationdb/metadata-foundationdb.c
Comment (nielspardon) [inline metadata-foundationdb/metadata-foundationdb.c]: should we prefix this with `g_`?

File: metadata/CMakeLists.txt
Comment (nielspardon) [inline metadata/CMakeLists.txt]: seems you are not using glib2 anymore so I guess this can be removed

File: metadata-foundationdb/metadata-foundationdb.c
Comment (nielspardon) [inline metadata-foundationdb/metadata-foundationdb.c]: maybe this should be called `metadata_load` to follow the same pattern of `<object>_<verb>` as with `metadata_persist` above

File: metadata-foundationdb/CMakeLists.txt
Comment (nielspardon) [inline metadata-foundationdb/CMakeLists.txt]: I think we need to revisit whether `${PROJECT_BINARY_DIR}` is the right value here for ´make install´.

**Review by nielspardon — CHANGES_REQUESTED** [review]
a couple of comments / changes

File: metadata-foundationdb/metadata-foundationdb.h
Comment (zoltan) [reply metadata-foundationdb/metadata-foundationdb.h]: good catch, thanks

File: metadata-foundationdb/metadata-foundationdb.c
Comment (zoltan) [reply metadata-foundationdb/metadata-foundationdb.c]: I originally took that prefix from some of the glib example code.

Looking up visibility rules in C, it seems like the best option is to declare it static - this should restrict visibility to the compilation unit, eg. metadata-foundationdb.c itself.

File: metadata/CMakeLists.txt
Comment (zoltan) [reply metadata/CMakeLists.txt]: indeed, as discussed, I've switched to pure dlopen() as for right now it works for us equally both.

File: metadata-foundationdb/CMakeLists.txt
Comment (zoltan) [reply metadata-foundationdb/CMakeLists.txt]: I'm open to suggestions. right now I never do make install.

File: metadata-foundationdb/metadata-foundationdb.c
Comment (zoltan) [reply metadata-foundationdb/metadata-foundationdb.c]: since the module is explicitly tied to metadata, we could get rid of the first part, and just leave the verb.

**Comment by zoltan** [comment]
the new changes should resolve all your requested changes. I've tested it with a local FDB server and it worked.

File: samples/metadata-sample.c
Comment (nielspardon) [inline samples/metadata-sample.c]: I would move the `Metadata *metadata = new_metadata();` back down after the `metadata_backend_initialize();` since we only need to start allocating memory if the backend initialization was successful and we have loaded the metadata backend successfully. Otherwise we are leaking some memory if we abort while setting up the backend.

File: metadata-foundationdb/CMakeLists.txt
Comment (nielspardon) [reply metadata-foundationdb/CMakeLists.txt]: what about `${CMAKE_INSTALL_LIBDIR}/hyperwarp` which would result in something like `/usr/local/lib/hyperwarp`?

**Review by nielspardon — CHANGES_REQUESTED** [review]
I think this is looking way better now. There's one small issue with possible memory leakage that should be fixed.

File: metadata-foundationdb/CMakeLists.txt
Comment (zoltan) [reply metadata-foundationdb/CMakeLists.txt]: sounds good

File: samples/metadata-sample.c
Comment (nielspardon) [inline samples/metadata-sample.c]: that's already better however I thought about calling it after `metadata_backend_initialize();` two lines further below

**Review by nielspardon — CHANGES_REQUESTED** [review]

File: samples/metadata-sample.c
Comment (zoltan) [reply samples/metadata-sample.c]: you're right. I've now improved the error handling. Please take a look.

**Review by nielspardon — APPROVED** [review]
LGTM if you now could squash the commits into one before merging :-)

**Pull request merged**
